### PR TITLE
Refactor impressions

### PR DESCRIPTION
**Pull request by maksimf**

### CODE REVIEW

File: lib/cache/senders/impressions_sender.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            Thread.new do
+              loop do
+                post_impressions
+
+                sleep(randomize_interval(@config.impressions_refresh_rate))
+              end
+            end
+          end
+
+          @config.logger.info('Started impressions service') unless ENV['SPLITCLIENT_ENV'] == 'test'
```
Comment (patricioe) [inline lib/cache/senders/impressions_sender.rb]: should we move ` @config.logger.info('Started impressions service')` between:
```ruby
Thread.new do
  @config.logger.info('Started impressions service') # <----- ?
  loop do
     ....
```

??

File: lib/cache/senders/impressions_sender.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      class ImpressionsSender
+        def initialize(impressions_repository, config, api_key)
+          @impressions_repository = impressions_repository
+          @config = config
+          @api_key = api_key
+        end
+
+        def call
+          # Disable impressions if @config.impressions_queue_size == -1
+          if @config.impressions_queue_size < 0
```
Comment (patricioe) [inline lib/cache/senders/impressions_sender.rb]: do we have a unit test that post impressions when it is disabled to make sure it doesn't blow?

**Comment by maksimf** [comment]
@patricioe Done, fixed

File: spec/engine_spec.rb [hunk 3]
```diff
@@ -272,6 +272,18 @@
...
+
+      context 'when impressions are disabled' do
+        let(:config) do
+          { logger: Logger.new('/dev/null'), cache_adapter: cache_adapter, impressions_queue_size: -1 }
+        end
+
+        it 'works when impressions are disabled' do
+          expect(subject.get_treatments('21', ["sample_feature", "beta_feature"])).to eq(
+            { sample_feature: 'off', beta_feature: 'off' }
+          )
```
Comment (patricioe) [inline spec/engine_spec.rb]: shouldn't we check that there are no impressions in the queue here?

File: lib/splitclient-rb/version.rb [hunk 4]
```diff
@@ -1,3 +1,3 @@
 module SplitIoClient
-  VERSION = '3.1.0-rc12'
+  VERSION = '3.2.0'
```
Comment (patricioe) [inline lib/splitclient-rb/version.rb]: Please don't change this here yet.

File: lib/engine/api/client.rb [hunk 5]
```diff
@@ -24,6 +37,15 @@ def api_client
           builder.adapter :net_http_persistent
         end
       end
+
+      def common_headers(api_key, config)
+        {
+          'Authorization' => "Bearer #{api_key}",
+          'SplitSDKVersion' => SplitIoClient::SplitFactory.sdk_version,
+          'SplitSDKMachineName' => config.machine_name,
+          'SplitSDKMachineIP' => config.machine_ip
```
Comment (patricioe) [inline lib/engine/api/client.rb]: do we need to cherry pick the referer commit here?

File: lib/cache/senders/impressions_sender.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+          if ENV['SPLITCLIENT_ENV'] == 'test'
+            post_impressions
+          else
+            Thread.new do
+              @config.logger.info('Starting impressions service') unless ENV['SPLITCLIENT_ENV'] == 'test'
+
+              loop do
+                post_impressions
+
+                sleep(randomize_interval(@config.impressions_refresh_rate))
```
Comment (patricioe) [inline lib/cache/senders/impressions_sender.rb]: do we need to cherry pick or rebase devleopment to make sure your Utilities class is used instead?

**Pull request merged**